[Broadlink] Initial contribution

AntonJansen opened the pull request:
# Title
- [Broadlink] A binding that can communicate with the Broadlink family of WiFi-addressable devices

# Description
The binding support the following devices:
- A1 environmental sensor
- RMx Remote IR blasters
- SPx WiFi-controlled power switches
- MPx WiFi-controlled multi-outlet power strips

This pull request is an update of the stalled pull request https://github.com/openhab/openhab-addons/pull/14582, updated for 4.2.x

The discussion for this pull request is:
https://community.openhab.org/t/broadlink-binding-4-1-0-4-2-0/154734

Signed-off-by: Anton Jansen [mail@antonjansen.com](mailto:mail@antonjansen.com)

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+    @Override
+    @SuppressWarnings("null")
+    public void dispose() {
+        logger.debug("Disposing of {}", getThing().getLabel());
```
```suggestion
```
This will already get logged in the event.log file if I am correct.
There is a small system load here even with debug off.
1. The string is created and built even if your not in debug mode.
2. The system checks if debug or above is active, another load.
3. If the log actually gets written you then hit another load which with slow SD cards can be high.
4. Garbage collection on the String that was created.

Sorry if this is obvious or you know it, but there are reasons why we like minimal logs unless something went wrong.

https://www.openhab.org/docs/developer/guidelines.html#f-logging

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+    public void dispose() {
+        logger.debug("Disposing of {}", getThing().getLabel());
+        if (refreshHandle != null && !refreshHandle.isDone()) {
+            logger.debug("Cancelling refresh task");
```
```suggestion
```
Same reasons as above. The next log line two lines down is fine to keep as it gives useful info to help users out.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+        logger.debug("Disposing of {}", getThing().getLabel());
+        if (refreshHandle != null && !refreshHandle.isDone()) {
+            logger.debug("Cancelling refresh task");
+            boolean cancelled = refreshHandle.cancel(true);
+            logger.debug("Cancellation successful: {}", cancelled);
```
```suggestion
            logger.debug("Cancellation successful: {}", refreshHandle.cancel(true));
```
Slightly more efficient and an example of how it will still run even if debug is disabled. Not necessary to make this change if you do not want to.

AntonJansen replied:
I don't like mixing control flow change and logging in one line. It reduces readability of the code imho. Hence, I suggest not to implement this change,

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    private boolean authenticate() {
+        logger.trace("Authenticating");
```
```suggestion
```
same reason

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+            authenticated = true;
+            return true;
+        } catch (Exception e) {
+            logger.warn("Authentication failed: ", e);
```
```suggestion
            logger.warn("Authentication failed: {}", e.getMessage());
```
Missing the brackets and also if you do not add the e.getMessage() you will get a comment in the report.html file.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            logger.trace("Refresh requested, updating item status ...");
```
```suggestion
```
same reason

Skinah reviewed (changes requested):
Please go through all uses of the logger and fix.

Also after you have built the binding look in the target folder where the jar appears.
Folder called target\code-analysis
open the file called report.html
Then try and fix as many as you can whilst you wait for more review comments.

Thanks for taking on the binding.

AntonJansen commented:
@Skinah the changes requested should be implemented. One thing I am struggling with is how the target\code-analysis is created. Is there a maven command that generates this locally, or can I only read it out of the CI report?

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/BroadlinkBindingConstants.java:
```diff
@@ -0,0 +1,110 @@
+    public static final ThingTypeUID THING_TYPE_RM4_PRO = new ThingTypeUID("broadlink", "rm4pro");
+    public static final ThingTypeUID THING_TYPE_A1 = new ThingTypeUID("broadlink", "a1");
+    public static final ThingTypeUID THING_TYPE_MP1 = new ThingTypeUID("broadlink", "mp1");
+    public static final ThingTypeUID THING_TYPE_MP1_1K3S2U = new ThingTypeUID("broadlink", "mp1_1k3s2u");
```
```suggestion
    public static final ThingTypeUID THING_TYPE_MP1_1K3S2U = new ThingTypeUID("broadlink", "mp1-1k3s2u");
```
Excellent that this is a constant, as it will make changing it far easier if all your code/classes uses THING_TYPE_MP1_1K3S2U in it.

https://www.openhab.org/docs/developer/guidelines.html#b-code-formatting-rules-style

lower-case-hyphen is what should be used for the thing type. I believe the others are fine to leave as is, but this uses the underscore, when the rule is to use hyphen. This kind of mix up really annoys the people that like textual config as a typo is easy to make and then hard to find so we have a rule. Try to have the thing type all lower and a single word if possible.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkA1Handler.java:
```diff
@@ -0,0 +1,65 @@
+            updateState("air", ModelMapper.getAirValue(decryptResponse[10]));
+            updateState("noise", ModelMapper.getNoiseValue(decryptResponse[12]));
+            return true;
+        } catch (Exception ex) {
+            logger.warn("Failed while getting device status", ex);
```
```suggestion
        } catch (Exception ex) {
            logger.warn("Failed while getting device status: {}", ex.getMessage());
```
This will be flagged in the code analysis report I mentioned in the target folder. It will pick up when you do not have an non equal amount of {}. Please read the report and fix as many as you can so the reviewers have less work, thanks.

jlaur replied:
No, the method has this signature:
```java
public void warn(String msg, Throwable t);
```

It logs the full exception stack trace. Usually this is most useful for **debug**/**trace** logs and a bit too much for **warn**, so your proposed change is good, but for other reasons. 🙂

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,293 @@
+            byte authRequest[] = buildMessage((byte) 0x65, BroadlinkProtocol.buildAuthenticationPayload(), -1);
+            byte response[] = sendAndReceiveDatagram(authRequest, "authentication");
+            if (response == null) {
+                logger.warn("response from device during authentication was null");
```
Is it possible to give a warning that guides the user what they need to change? I have not read all code so if this is provided by the thing going offline with a helpful message then that is enough.

AntonJansen replied:
Added actionable guidance text to check the MAC address of the device is configured correctly.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkRemoteHandler.java:
```diff
@@ -0,0 +1,194 @@
+                logger.info("END LAST LEARNT CODE ({} characters)", hexString.length());
+            }
+        } catch (IOException e) {
+            logger.warn("Exception while attempting to check learnt code", e);
```
same here

AntonJansen replied:
Fixed

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/i18n/broadlink_xx.properties:
```diff
@@ -0,0 +1,21 @@
+# FIXME: please substitute the xx with a proper locale, ie. de
+# FIXME: please do not add the file to the repo if you add or change no content
```
You need to follow this text, either delete the file if you change no content, or you need to rename it to _en.properties for english. Once you have done it, remove the fixme

AntonJansen replied:
Re-generated the internationalization file.

Skinah reviewed (changes requested):
You need to run the mvn command here to create the translation files so that people can translate the binding into multiple languages. Command is here as well as background info:

https://www.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file

When you change the XML files ie adding new channel, changing thing types etc.. you will need to re-run this command to re-create the files.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/mp-types.xml:
```diff
@@ -0,0 +1,58 @@
+	</thing-type>
+	<thing-type id="mp2">
+		<label>Broadlink MP2</label>
+		<description>The Broadlink MP2 is a smart socket with Wi-Fi connectivity. It does not support switching by individual
```
```suggestion
		<description>The Broadlink MP2 is a smart socket with Wi-Fi connectivity. It does not support switching individual
```

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/rm-types.xml:
```diff
@@ -0,0 +1,64 @@
+	</thing-type>
+	<thing-type id="rm3q">
+		<label>Broadlink RM3 v44057</label>
+		<description>The Broadlink RM 3/Mini is a Wi-Fi IR transmitter running v44057 firmware.</description>
```
This sounds interesting, can you explain more as to why an exact firmware is mentioned in the description and label? The labels get used when items are auto created by `create equipement from thing` and this seems very odd to include the firmware here. Is there a good reason, if so please explain as any reviewer is going to ask this.

AntonJansen replied:
It seems that there are 2 versions of the product with the same product name, but with a different firmware for different markets. The v44057 is a firmware specific to the Chinese market, but through grey imports have landed in the US / EU. However, in this market the same product with the same name, but with a different firmware is being sold. 

The v44057 firmware seems to have issues responding on status updates. Hence, a workaround was created.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,46 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
+	<thing-type id="sample">
+		<!-- Comment this in for Things which need a Bridge to function.
+			<supported-bridge-type-refs>
+			<bridge-type-ref id="BroadlinkBridge" />
+			</supported-bridge-type-refs>
+		-->
+
+		<label>Broadlink Binding Thing</label>
+		<description>Sample thing for Broadlink Binding</description>
+
+		<channels>
+			<channel id="channel1" typeId="sample-channel"/>
```
Looks like this whole file needs to be deleted?

AntonJansen replied:
Yup, this seems to be part of some initial boiler plate code that was never included in the binding in the end. Removed the associated files.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,53 @@
+		<parameter name="staticIp" type="boolean">
+			<context>static-ip</context>
+			<label>Static IP</label>
+			<description>Will the device always be given this network address?</description>
```
Interesting config to have, can you provide some background why this is used or needed? Usually I just make sure that the address config supports both IP and also hostnames, so a hostname can be used if it is not static.

AntonJansen replied:
I have no idea why this is in here. Since parts of this code is already 5+ years old, I am a bit hesitant to remove functionality, as people might want to use it.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/discovery/DiscoveryProtocol.java:
```diff
@@ -0,0 +1,84 @@
+            BroadlinkSocket.registerListener(listener, logger);
+            DiscoveryProtocol.discoverDevices(logger);
+            DiscoveryProtocol.waitUntilEnded(timeoutMillis, logger);
+            logger.warn("Ended Broadlink device scan...");
```
```suggestion
```
Not sure if this warrants a warning as this is normal program flow?

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/discovery/DiscoveryProtocol.java:
```diff
@@ -0,0 +1,84 @@
+        try {
+            logger.warn("Broadlink device scan waiting for {} ms to complete ...", discoveryTimeoutMillis);
+            Thread.sleep(discoveryTimeoutMillis);
+            logger.warn("Device scan: wait complete ...");
```
When does this occur? should this be debug level if it only happens when it times out, or is it a serious thing that needs a warn?

AntonJansen replied:
Cleaned up the logging for this part.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/discovery/DiscoveryProtocol.java:
```diff
@@ -0,0 +1,84 @@
+
+    private static void waitUntilEnded(long discoveryTimeoutMillis, Logger logger) {
+        try {
+            logger.warn("Broadlink device scan waiting for {} ms to complete ...", discoveryTimeoutMillis);
```
Can it be removed? warn seems wrong here, debug or trace levels unless it actually causes an issue.
Please can you do a FIND on your files for `logger.warn` and review yourself if they need to be removed, or changed to another level like debug or trace if they create a lot of lines?

Guide for logging is here and covers when each level should be used.

https://www.openhab.org/docs/developer/guidelines.html#f-logging

AntonJansen replied:
Cleaned up all the logger.warn statements.

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/BroadlinkSocket.java:
```diff
@@ -0,0 +1,160 @@
+                    logger.warn("Error while receiving", e);
+                }
+            }
+            logger.debug("Receiver thread ended");
```
```suggestion
```

AntonJansen replied:
Fixed

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/BroadlinkSocket.java:
```diff
@@ -0,0 +1,160 @@
+                }
+            } catch (IOException e) {
+                if (!isInterrupted()) {
+                    logger.warn("Error while receiving", e);
```
\target\code-analysis\report.html

AntonJansen replied:
Fixed

Skinah commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/RetryableSocket.java:
```diff
@@ -0,0 +1,116 @@
+            socket.send(sendPacket);
+            return true;
+        } catch (IOException e) {
+            logger.warn("IO error during UDP command sending {}:", purpose, e);
```
\target\code-analysis\report.html

AntonJansen replied:
Fixed

Skinah commented:
Most of the basics look really good, if you can post when you have fixed as much as you can from \target\code-analysis\report.html I can take another look and then call in a reviewer that knows more then me to have a look @hmerk (showed interest in this binding).  If you do not agree with any of my suggestions please feel free to say so as I do not know the code and your reasons like you do.

One thing to note is that the constant class, all the handlers and BroadlinkDeviceConfiguration.java are not under the internal package/folder. I believe they need to be, just not sure of the reason why. Having them in packages/folders under `internal` is fine like the rest have been done.

AntonJansen commented:
@Skinah currently I am using "mvn verify" to generate the code-analysis\report.html, is that correct? As I do not see any issues in the last report, so I wonder were your comments come from / why I do not see them reflected in the report.

I will go over your comments the coming days.

Skinah commented:
'mvn clean install' in the bindings folder, then look in the target folder where the jar is built. How are you building the jar?

AntonJansen commented:
I have been using mvn package thus far or the automated maven execution of Eclipse.

hmerk commented:
@Skinah I will have a look but this might take some time as I am preparing for FOSDEM next weekend in Brussels.

AntonJansen commented:
@Skinah  I have implemented all your requested changes. @hmerk enjoy your conference! I will be  travelling 5/2 -10/2 and not have time to work on this, so the perfect time for some reviews :-)

AntonJansen commented:
I forgot the internationalization part. Will try to fix it in the coming days.

AntonJansen commented:
@hmerk do you still have time and/or intent to review this?

hmerk commented:
@AntonJansen Of course I intent to review your PR, it is just a matter off time as I sm moving to a new flat atm.
Hope to find some timeslots in the next weeks.

AntonJansen commented:
@hmerk any update from your side?
I still have a branch to merge (rm4pros) that will bring in RF support and a way to create new IR/RF commands through the UI. Still need to clean up the code.

AntonJansen commented:
@hmerk any update from your side?
I still have a branch to merge (rm4pros) that will bring in RF support and a way to create new IR/RF commands through the UI. Still need to clean up the code.

jlaur commented on bundles/org.openhab.binding.broadlink/src/main/resources/broadlink.png:
The logo should be provided in a PR towards the openhab-docs repository. See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website

AntonJansen replied:
@jlaur it says "After your pull request has been merged and the next openHAB version is released, your add-on will be available in the addons search on the openHAB website with a default logo". I'm still waiting on the merge ....

jlaur replied:
That's correct. You can either create a draft PR with the logo already, or you can wait with that until the PR is merged. My comment is towards this PR: The logo should not be included here.

AntonJansen replied:
Right, but where should I store it in the meantime? I guess I will move it to another branch.

jlaur replied:
You can store it in a draft PR towards the openhab-docs repository, or you can store it on your local drive or anywhere else you wish. 😉 If/when you create a PR, please provide a link to this PR, so the maintainers can easily check the status.

AntonJansen replied:
@jlaur I have removed the logo as requested. A pull request (https://github.com/openhab/openhab-docs/pull/2277) has been created to update the logo in the documentation.

As @hmerk has indicated not to have time to review the binding, can you arrange another reviewer or pick it up yourself to bring it to the finish line?

hmerk commented:
> @hmerk any update from your side? I still have a branch to merge (rm4pros) that will bring in RF support and a way to create new IR/RF commands through the UI. Still need to clean up the code.

Sorry for being quiet lately, but life changed a lot over at my side. Unfortunately, I will not be able to review your code for the next couple of month. But this should not be a big issue, as I am no maintainer and therefore not able to merge approve your PR.

AntonJansen commented:
@Skinah  can you take another look at this?

AntonJansen commented:
Hi all, the improvements mentioned before (RF support, Openhab storage service) are now part of the PR. Apart from the DCO issue, the binding is now done and ready to be merged from our point of view.

lsiepel commented:
> Hi all, the improvements mentioned before (RF support, Openhab storage service) are now part of the PR. Apart from the DCO issue, the binding is now done and ready to be merged from our point of view.

Looking into some PR's i see this one. Looks liek a great binding.  Might fully review it later, a quickscan showed that some channel id's are not according to https://www.openhab.org/docs/developer/guidelines.html#naming-convention

You might also do some self-review to speed up the proces: https://github.com/openhab/openhab-addons/wiki/Review-Checklist

lsiepel commented:
@AntonJansen are you able to proceed ?

AntonJansen commented:
@lsiepel, I do not get a response thus far from Ricardo Larranaga to fix his git commits comments as to pass the DCO check. I'm just back from an extended holiday. Will spend some time the coming days fixing the channel ids and do the self-review.

AntonJansen commented:
@lsiepel I have made the changes to the channels names as you suggested. I also went through the checklist and did not find any issues.

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,90 @@
+		<state readOnly="true"></state>
+	</channel-type>
+	<channel-type id="air">
+		<item-type>String</item-type>
```
Is this a constant set of known strings? If so it would be nice to have those listed in an option tag. (same for other string channel id's)

AntonJansen replied:
I have no idea, as I do not own these devices.

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,90 @@
+		<category>Temperature</category>
+		<state pattern="%.1f %unit%" readOnly="true"/>
+	</channel-type>
+	<channel-type id="humidity">
+		<item-type>Number:Dimensionless</item-type>
+		<label>Humidity</label>
+		<category>Humidity</category>
+		<state readOnly="true" min="0" max="100" pattern="%.2f %unit%"/>
+	</channel-type>
```
Think this can be removed as there is a system variant for this: system.atmospheric-humidity

AntonJansen replied:
Done

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
binding.xml is no longer used since openHAB 4.x Please remove.

AntonJansen replied:
Done

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/RetryableSocket.java:
```diff
@@ -0,0 +1,116 @@
+        return null;
+    }
+
+    @SuppressWarnings("null")
```
Remove null suppressions

AntonJansen replied:
Fixed

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/RetryableSocket.java:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    @SuppressWarnings("null")
```
Remove null suppressions

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/RetryableSocket.java:
```diff
@@ -0,0 +1,116 @@
+        return null;
+    }
+
+    @SuppressWarnings("null")
```
Remove null suppressions

AntonJansen replied:
Done

AntonJansen replied:
It should be gone now.

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/NetworkTrafficObserver.java:
```diff
@@ -0,0 +1,28 @@
+@NonNullByDefault
+public interface NetworkTrafficObserver {
+
+    public void onCommandSent(byte command);
```
Add javadoc to all interface methods

AntonJansen replied:
Done

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/BroadlinkSocketListener.java:
```diff
@@ -0,0 +1,28 @@
+@NonNullByDefault
+public interface BroadlinkSocketListener {
+
+    public abstract void onDataReceived(String remoteAddress, int remotePort, String remoteMAC,
```
Add java doc to all interface methods.

lsiepel reviewed (changes requested):
Thanks for this contribution! Just reviewed 2/3 and start to repeat similar comments so i stop now and let you look at the current comments. Ping me whenever you are ready.

AntonJansen commented:
I hope to have some time this weekend to work on the requested changes.

AntonJansen commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,184 @@
+In this version of the binding, the IR codes are store in `$OPENHAB_USERDATA/jsondb/broadlink_ir.json` and for the RM Pro series of devices
+the RF codes are store in `$OPENHAB_USERDATA/jsondb/broadlink_rf.json`
+
+In previous versions of this binding, the file used the [Java Properties File format](https://en.wikipedia.org/wiki/.properties) and was stored in the `<OPENHAB_CONF>/transform` folder.
```
A lot of people come from a very old version of the binding that was working up to openHAB version 3.3. They do not want to have to re-learn their commands, as it is sometimes impractical todo so. Hence, the documentation explains how to migrate at the end of the documentation.

lsiepel replied:
I totally understand, also think this should be more specific:
```suggestion
Before openHAB version 4.3.0, the file used the [Java Properties File format](https://en.wikipedia.org/wiki/.properties) and was stored in the `<OPENHAB_CONF>/transform` folder.
```

AntonJansen commented on bundles/org.openhab.binding.broadlink/README.md:
````diff
@@ -0,0 +1,184 @@
+By default, the file name was `broadlink.map` for the IR codes, but could be changed using the `mapFile` setting. In similar fashion, the RM pro models
+stored the RF codes in the `broadlinkrf.map` file. 
+
+Here is a map file example of the previous file format:
+
+```
+TV_POWER=26008c0092961039103a1039101510151014101510151039103a10391015101411141015101510141139101510141114101510151014103a10141139103911391037123a10391000060092961039103911391014111410151015101411391039103a101411141015101510141015103911141015101510141015101510391015103911391039103a1039103911000d05000000000000000000000000
+heatpump_off=2600760069380D0C0D0C0D290D0C0D290D0C0D0C0D0C0D290D290D0C0D0C0D0C0D290D290D0C0D0C0D0C0D0C0D0C0D0C0D0C0D0C0D0C0D0C0D0C0D0C0D0C0D290D0C0D0C0D0C0D0C0D0C0D0C0D0C0D290D0C0D0C0D0C0D0C0D290D0C0D0C0D0C0D0C0D0C0D0C0D290D0C0D290D290D290D290D290D290E0002900000
+```
````
As the binding was never part of the official distribution, it seems to me it should not be part of the upgrade messaging. Instead, I think the end of the binding documentation is a good place to store this.

AntonJansen replied:
Added clarification text in the documentation that this is about legacy migration.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/ModelMapper.java:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    private static final StringType[] AIR_VALUES = { new StringType("PERFECT"), new StringType("GOOD"),
+            new StringType("NORMAL"), new StringType("BAD") };
+
+    public static StringType getAirValue(byte b) {
+        return lookup(AIR_VALUES, b);
+    }
+
+    private static final StringType[] LIGHT_VALUES = { new StringType("DARK"), new StringType("DIM"),
+            new StringType("NORMAL"), new StringType("BRIGHT") };
+
+    public static StringType getLightValue(byte b) {
+        return lookup(LIGHT_VALUES, b);
+    }
+
+    private static final StringType[] NOISE_VALUES = { new StringType("QUIET"), new StringType("NORMAL"),
+            new StringType("NOISY"), new StringType("EXTREME") };
```
I have introduced an enum for this.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/discovery/BroadlinkDiscoveryService.java:
```diff
@@ -0,0 +1,111 @@
+    @Override
+    public void onDiscoveryFinished() {
+        logger.info("Discovery complete. Found {} Broadlink devices", foundCount);
+    }
```
Need to implement this method to fullfill the requirements of the abstract interface.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/BroadlinkMappingService.java:
```diff
@@ -0,0 +1,207 @@
+    public @Nullable String lookupCode(String command, String codeType) {
+        String response;
+        switch (codeType) {
+            case "IR":
```
Codetype is now a enum.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/NetworkUtils.java:
```diff
@@ -0,0 +1,114 @@
+
+    public static boolean hostAvailabilityCheck(@Nullable String host, int timeout, Logger logger) {
+        if (host == null) {
+            logger.warn("Can't check availability of a null host");
```
The method is called during initialization of the binding.  I will add a check for the configuration at the start.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/NetworkUtils.java:
```diff
@@ -0,0 +1,114 @@
+            InetAddress address = InetAddress.getByName(host);
+            return address.isReachable(timeout);
+        } catch (Exception e) {
+            logger.warn("Exception while trying to determine reachability of {}: {}", host, e.getMessage());
```
Fixed

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/NetworkUtils.java:
```diff
@@ -0,0 +1,114 @@
+        }
+        try {
+            InetAddress address = InetAddress.getByName(host);
+            return address.isReachable(timeout);
```
This utility is used to determine / update the status of the thing. It is called by updateItemStatus() in the BroadlinkBaseThingHandler, to ensure we set the correct thing state.

lsiepel replied:
My question remains the same. This is an out of context check and therefore not so usefull. Let's sey there is a sequence of events:
1. the bindings inits or performans an update or other event that `updateItemStatus()` is called. 
2. Everytime `updateItemStatus()` is called the host is checked for reachability
3. The host may become unavailable between this check and the actuall API-call retrieving the data. Making it essential to also handle that exception at the actuall API-call. 

2 is overhead and 3 shows this check has no use as it is not safe.
Also see https://github.com/openhab/openhab-addons/pull/16273/files#r1659249350

AntonJansen replied:
After looking at the situation, I think the binding is trying to do the following. When there is an event or on a regular basis the uipdateItemStatus() method is called. It first checks for the reachability of the thing. If the host is not reachable interesting things happen. If the IP is defined as static it does nothing. However, if it is dynamic it starts a thread to discover if the device is on a new IP address. I think this is the primary reason for the reachability check. Ideally, this should be rewritten to a more clean design. However, this is complicated as it breaks the structure of the current implementation due to inheritance relationships.

To be honest, for me this is not a show stopper and rather a nice to have then a must have. The main reason is that in most cases the overhead is not significant.

lsiepel replied:
The mechanism is not wrong, it is the implementation that is far from optimal. AS stated above, there are unhandlked race conditions and it is called excessive.
It can be fixed by adjusting it a little bit. 
Should not be that much work. I allready created an example pseudo code  here https://github.com/openhab/openhab-addons/pull/16273/files/1cbc3968e5e955e21844f35431a1aaec7f7a33b5#r1659249350

that can be extended with a catch block specific for when the host is not reachable: (have to check the right thrown exception classes, used it out of my head) 
```
try {
    getStatusFromDevice();
    updateStatus(ThingStatus.ONLINE);
} catch (NoRouteToHostEception | HostNotFoundException ex) {
    // this block should be hit whenever the host is not reachable. Here you can run recovery code:
            if (thingConfig.isStaticIp()) {
                    forceOffline(ThingStatusDetail.NONE, "Couldn't find statically-IP-addressed device");
            } else {
                DeviceRediscoveryAgent dra = new DeviceRediscoveryAgent(thingConfig, this);
                dra.attemptRediscovery();
                logger.debug("Asynchronous dynamic IP device search initiated...");
            }
    }
} catch (CustomAuthenticationException ex) {
    //update Thing state to OFFLINE set detail to CONFIGURATION_ERROR nad message to "Credentials are invalid";
    }
} catch (CustomCommuncationException ex) {
    //update Thing state to OFFLINE set detail to COMMUNICATION_ERROR and message to "Fatal timeout after 3 retries"
    }
}
```

AntonJansen replied:
The code has been updated.

lsiepel commented:
Just a check; not all comments are fixed, the ones behind the 'show more' button are not commented on.

AntonJansen commented:
Yes, I believe there are still 27 comments to go through.On 12 Jul 2024, at 22:53, lsiepel ***@***.***> wrote:﻿
Just a check; not all comments are fixed, the ones behind the 'show more' button are not commented on.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/NetworkUtils.java:
```diff
@@ -0,0 +1,114 @@
+        return false;
+    }
+
+    public static @Nullable InetAddress findNonLoopbackAddress() throws SocketException {
```
Done :-)

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/BroadlinkBindingConstants.java:
```diff
@@ -0,0 +1,118 @@
+    public static final String BROADLINK_AUTH_KEY = "097628343fe99e23765c1513accf8b02";
+    public static final String BROADLINK_IV = "562e17996d093d28ddb3ba695a2e6f58";
+
+    public static final String COMMAND_CHANNEL = "command";
+    public static final String LEARNING_CONTROL_CHANNEL = "learning-control";
+    public static final String RF_COMMAND_CHANNEL = "rfcommand";
+    public static final String RF_LEARNING_CONTROL_CHANNEL = "learning-rf-control";
```
Thank you for addressing this. You can leave it as it is now, but somehow i mixed them up, my preference is to have the prefix variant. Anyway, lets continue, they are consistent, that is more important.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/Utils.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@NonNullByDefault
+public class Utils {
+    public static boolean isOnline(Thing thing) {
```
Fixed

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/config/BroadlinkDeviceConfiguration.java:
```diff
@@ -0,0 +1,128 @@
+    private int deviceType;
+
+    public BroadlinkDeviceConfiguration() {
+        ipAddress = "";
+        staticIp = true;
+        macAddress = "";
+        macAddressBytes = new byte[0];
+        pollingInterval = 30;
+        nameOfCommandToLearn = "DEVICE_ON";
+        ignoreFailedUpdates = false;
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/config/BroadlinkDeviceConfiguration.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    public boolean isStaticIp() {
+        return staticIp;
```
To be honest, I have no idea. This code is very old. I have just taken it up to clean out and update it to openhab 4.x

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/discovery/DeviceRediscoveryListener.java:
```diff
@@ -0,0 +1,27 @@
+ */
+@NonNullByDefault
+public interface DeviceRediscoveryListener {
+    void onDeviceRediscovered(String newIpAddress);
```
Fixed

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/discovery/DiscoveryFinishedListener.java:
```diff
@@ -0,0 +1,25 @@
+ */
+@NonNullByDefault
+public interface DiscoveryFinishedListener {
+    void onDiscoveryFinished();
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkA1Handler.java:
```diff
@@ -0,0 +1,65 @@
+
+            updateTemperature(temperature);
+            updateHumidity((decryptResponse[6] * 10 + decryptResponse[7]) / 10D);
+            updateState("light", ModelMapper.getLightValue(decryptResponse[8]));
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+
+    @Override
+    public void initialize() {
+        this.thingConfig = getConfigAs(BroadlinkDeviceConfiguration.class);
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+                public void run() {
+                    updateItemStatus();
+                }
+            }, 1L, thingConfig.getPollingInterval(), TimeUnit.SECONDS);
```
This sadly does not work.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+    public void initialize() {
+        this.thingConfig = getConfigAs(BroadlinkDeviceConfiguration.class);
+        count = (new Random()).nextInt(65535);
+
```
Validation added

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    @Override
+    @SuppressWarnings("null")
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+    @Override
+    @SuppressWarnings("null")
+    public void dispose() {
+        if (refreshHandle != null && !refreshHandle.isDone()) {
+            boolean cancelled = refreshHandle.cancel(true);
+            logger.debug("Cancellation successful: {}", cancelled);
+        }
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+        }
+    }
+
+    @SuppressWarnings("null")
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkRemoteHandler.java:
```diff
@@ -0,0 +1,465 @@
+        return Utils.slice(decryptedResponse, 4, decryptedResponse.length);
+    }
+
+    @SuppressWarnings("null")
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/sp-types.xml:
```diff
@@ -0,0 +1,57 @@
+		<label>Broadlink SP1</label>
+		<description>The Broadlink SP1 is a Wi-Fi smart socket.</description>
+		<channels>
+			<channel id="power-on" typeId="power-on"/>
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/sp-types.xml:
```diff
@@ -0,0 +1,57 @@
+		<label>Broadlink SP2</label>
+		<description>The Broadlink SP2 is a Wi-Fi smart socket.</description>
+		<channels>
+			<channel id="power-on" typeId="power-on"/>
```
Done

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/thing/sp-types.xml:
```diff
@@ -0,0 +1,57 @@
+		<label>Broadlink SP2s</label>
+		<description>The Broadlink SP2s is a Wi-Fi smart socket with power consumption measurement.</description>
+		<channels>
+			<channel id="power-on" typeId="power-on"/>
```
Done

lsiepel reviewed (changes requested):
When commenting on your update i noticed i had some pending comments.. Flushed them now.
As you replied on one of the comments, and there are more (4 or 5) left open, i like to be sure you are aware of them as they are burried under all the load more buttons.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+            authenticated = true;
+            return true;
+        } catch (Exception e) {
+            logger.warn("Authentication failed: {}", e.getMessage());
+            return false;
```
Exception added

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+        return true;
+    }
+
+    public void updateItemStatus() {
```
This would require a significant rewrite, as currently the inheritance structure is used to inherit both aspects.

AntonJansen replied:
Refactoring done.

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
````diff
@@ -0,0 +1,189 @@
+
+The code shown below is a Python script that can be used to convert from the old format to the new one:
+
+```
````
```suggestion
```python
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
````diff
@@ -0,0 +1,189 @@
+    print(json.dumps(result,indent=2))
+```
+
+
````
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+As an advantage, the files are now backed up by openHAB, which is more practical for migrations, data robustness, etc. having the storage of the codes handled by openHAB also provides uniformity in where the files are stored.
+
+With the change of the storage mechanism, the files are also changing format, and codes are now stored in json. 
+As an example, a file with the commands shown in the previous example would look like this:
```
```suggestion
With the change of the storage mechanism, the files are also changing format, and codes are now stored in json.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+To send either code, the string `TV_POWER` or `heatpump_off` must be sent to the `command` channel for the device.
+For RF, the `rfcommand` channel is used. 
+
+Storage of codes is handled by openHAB. The map files are stored in the $OPENHAB_USERDATA/jsondb directory. 
```
```suggestion
Storage of codes is handled by openHAB. The map files are stored in the $OPENHAB_USERDATA/jsondb directory.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+
+In previous versions of this binding, the file used the [Java Properties File format](https://en.wikipedia.org/wiki/.properties) and was stored in the `<OPENHAB_CONF>/transform` folder.
+By default, the file name was `broadlink.map` for the IR codes, but could be changed using the `mapFile` setting. 
+In similar fashion, the RM pro models stored the RF codes in the `broadlinkrf.map` file. 
```
```suggestion
In similar fashion, the RM pro models stored the RF codes in the `broadlinkrf.map` file.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+IR codes are store in `$OPENHAB_USERDATA/jsondb/broadlink_ir.json` and for the RM Pro series of devices the RF codes are store in `$OPENHAB_USERDATA/jsondb/broadlink_rf.json`
+
+In previous versions of this binding, the file used the [Java Properties File format](https://en.wikipedia.org/wiki/.properties) and was stored in the `<OPENHAB_CONF>/transform` folder.
+By default, the file name was `broadlink.map` for the IR codes, but could be changed using the `mapFile` setting. 
```
```suggestion
By default, the file name was `broadlink.map` for the IR codes, but could be changed using the `mapFile` setting.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+## Migrating legacy map file
+
+Up to openHAB version 3.3, there was a previous version of this binding that was not part of the openHAB distribution. 
+It stored the IR/RF commands in a different place and a different format. 
```
```suggestion
It stored the IR/RF commands in a different place and a different format.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+
+## Migrating legacy map file
+
+Up to openHAB version 3.3, there was a previous version of this binding that was not part of the openHAB distribution. 
```
```suggestion
Up to openHAB version 3.3, there was a previous version of this binding that was not part of the openHAB distribution.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
````diff
@@ -0,0 +1,189 @@
+
+Items file example; `sockets.items`:
+
+```
````
```suggestion
```java
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+4. Click the item, and click the rectangular area that is marked NULL
+5. In the pop-up menu that appears, select *Delete IR command* for IR or *Delete RF command* for RF
+
+*VERY IMPORTANT NOTE: Unlike the previous binding, writing the codes into the files is handled by openHAB. While it is possible to create a file externally, copy it in the proper location and use it as a remote codes database (As it is done in the case of Remote codes file migration) IT IS STRONGLY DISCOURAGED to modify the file while the binding is acive. Please make sure the binding is stopped  ...
```
```suggestion
*VERY IMPORTANT NOTE: As of openHAB version 4.3.0, writing the codes into the files is handled by openHAB. While it is possible to create a file externally, copy it in the proper location and use it as a remote codes database (As it is done in the case of Remote codes file migration) IT IS STRONGLY DISCOURAGED to modify the file while the binding is acive. Please make sure the binding is stopped before you modify a remote codes file manually. Also, have the following things in mind:*
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+10. If no succes, the channel will be named "NULL". Inspect the `openhab.log` file on your openHAB server for any issues
+11. Check and save the IR/RF command by clicking the item once more and select "Check and save command". 
+12. Keep pressing the remote control with the command to check and save
+12. If succesfull, the channel will change name to the command saved
```
```suggestion
13. If succesfull, the channel will change name to the command saved
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+11. Check and save the IR/RF command by clicking the item once more and select "Check and save command". 
+12. Keep pressing the remote control with the command to check and save
+12. If succesfull, the channel will change name to the command saved
+13. If no succes, the channel be named "NULL", restart from step 3.
```
```suggestion
14. If no succes, the channel be named "NULL", restart from step 3.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+8. *The LED on your RM device will extinguish once it has identified the command*
+9. If the command has been identified succesfully, the channel will have changed it name to "Learn command" or *RF command learnt*
+10. If no succes, the channel will be named "NULL". Inspect the `openhab.log` file on your openHAB server for any issues
+11. Check and save the IR/RF command by clicking the item once more and select "Check and save command". 
```
```suggestion
11. Check and save the IR/RF command by clicking the item once more and select "Check and save command".
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+
+## Learning Remote Codes
+
+To obtain the command codes, you can get this binding to put your Broadlink RMx device into "learn mode" and then ask it for the code it learnt. 
```
```suggestion
To obtain the command codes, you can get this binding to put your Broadlink RMx device into "learn mode" and then ask it for the code it learnt.
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+| command           | all RMx                  | String               | IR Command code to transmit                     |
+| learning-control  | all RMx                  | String               | Learn mode command channel (see below)          |
+
+
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,189 @@
+| macAddress          | String  |               | The device's MAC Address                                                          |
+| pollingInterval     | Integer | 30            | The interval in seconds for polling the status of the device                      |
+| nameOfCommandToLearn| String  | DEVICE_ON     | The name of the IR or RF command to learn when using the learn command channel    |
+| ignoreFailedUpdates | Boolean | false         | If enabled, failed status requests won't put the device `OFFLINE`                |
```
This is not used anywhere.
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,293 @@
+    // model. Return false if something went wrong that requires
+    // a change in the device's online state
+    protected void getStatusFromDevice() throws IOException, BroadlinkException {
+        NetworkUtils.hostAvailabilityCheck(thingConfig.getIpAddress(), 3000, logger);
```
This would make the NetworkUtils.hostAvailabilityCheck unnecesairy.
```suggestion
        InetAddress address = InetAddress.getByName(thingConfig.getIpAddress());
        if (!address.isReachable(3000)) {
            throw new BroadlinkHostNotReachableException("Cannot reach " + thingConfig.getIpAddress());
        }
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/NetworkUtils.java:
```diff
@@ -0,0 +1,153 @@
+     * @throws SocketException thrown when no socket can be opened.
+     */
+    public static @Nullable InetAddress findNonLoopbackAddress() throws SocketException {
+        Enumeration<NetworkInterface> ifaces = NetworkInterface.getNetworkInterfaces();
+        while (ifaces.hasMoreElements()) {
+            NetworkInterface iface = ifaces.nextElement();
+            Enumeration<InetAddress> inetAddrs = iface.getInetAddresses();
+            while (inetAddrs.hasMoreElements()) {
+                InetAddress inetAddr = inetAddrs.nextElement();
+                if (inetAddr.isLoopbackAddress()) {
+                    continue; /* Loop/switch isn't completed */
+                }
+
+                if (inetAddr.isSiteLocalAddress()) {
+                    return inetAddr;
+                }
+            }
+        }
+
+        return null;
```
Simplify by making use of the framework available API's

```suggestion
        for (InetAddress address : NetUtil.getAllInterfaceAddresses().stream().filter(a->a.getAddress() instanceof Inet4Address).map(a->a.getAddress()).toList()) {
            if (address.isSiteLocalAddress()) {
                return address;
            }
        }
        return null;
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkAuthenticationException.java:
```diff
@@ -0,0 +1,32 @@
+ */
+@NonNullByDefault
+public class BroadlinkAuthenticationException extends BroadlinkException {
+
```
```suggestion

        private static final long serialVersionUID = 6332210773192650617L;
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkException.java:
```diff
@@ -0,0 +1,32 @@
+ */
+@NonNullByDefault
+public class BroadlinkException extends Exception {
+
```
```suggestion

        private static final long serialVersionUID = 1L;
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkHostNotReachableException.java:
```diff
@@ -0,0 +1,28 @@
+ */
+@NonNullByDefault
+public class BroadlinkHostNotReachableException extends BroadlinkException {
+
```
```suggestion

        private static final long serialVersionUID = 1L;
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkRemoteHandler.java:
```diff
@@ -0,0 +1,496 @@
+
+    @Override
+    public void dispose() {
+        if (this.mappingService != null) {
+            this.mappingService.dispose();
```
Please fix all compiler warnings.
```suggestion
        BroadlinkMappingService mappingService = this.mappingService;
        if (mappingService != null) {
            mappingService.dispose();
            this.mappingSerivce = null;
```

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkRemoteHandler.java:
```diff
@@ -0,0 +1,496 @@
+                    message = "saved";
+                }
+                if (cmdLabel != null) {
+                    logger.info("BEGIN LAST LEARNT CODE");
+                    logger.info("{}", hexString);
+                    logger.info("END LAST LEARNT CODE ({} characters)", hexString.length());
```
In the past people had to copy & paste these codes into a mapping file. However, this is no longer required any more and it can be done through the UI instead.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/RetryableSocket.java:
```diff
@@ -0,0 +1,115 @@
+    @SuppressWarnings("null")
+    private @Nullable DatagramPacket receiveDatagram(String purpose, DatagramPacket receivePacket) {
+        try {
+            if (socket == null) {
```
Oopsie, they are gone now.

AntonJansen commented on bundles/org.openhab.binding.broadlink/src/main/resources/OH-INF/config/rm-config.xml:
```diff
@@ -0,0 +1,59 @@
+			<default>DEVICE_ON</default>
+			<required>true</required>
+		</parameter>
+		<parameter name="ignoreFailedUpdates" type="boolean">
+			<context>ignore-failed-updates</context>
+			<label>Ignore Failed Updates</label>
+			<description>Should failed status requests force the device offline?</description>
+			<required>true</required>
+			<advanced>true</advanced>
+			<default>false</default>
+		</parameter>
```
We removed this option when refactoring the updateItemStatus.

lsiepel commented:
I guess the commits wher emade though the Github web interface. When not properly configured, this causes the DCO to fail because of missing signoff. Unfortunately the suggestions had some whitespace issues so spotless is also complaining.

Hopefully you manage to rewrite the commit messages with a signoff and fix the spotless issue. I would also ask you to have a test run after all these change have been applied. It will not be the first and last time my suggestions lead to regression.

AntonJansen commented:
I should have addressed all your comments now. The main thing left is the DCO, which is due to Ricardo Larranaga, which did not use the proper sign-off. I have been trying to reach him the last 6 months, but no luck thus far.

lsiepel commented:
> @lsiepel I should have addressed all your comments now. The main thing left is the DCO, which is due to Ricardo Larranaga, which did not use the proper sign-off. I have been trying to reach him the last 6 months, but no luck thus far.

I’ll try to have another look next week.

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/NetworkUtils.java:
```diff
@@ -0,0 +1,143 @@
+     * @param to port number of the end of the range
+     * @return number of the available port
+     */
+    public static int nextFreePort(InetAddress host, int from, int to) {
```
You might want to add check `from < to`

Another issue is what if there is no free port, this method will end up in and endless loop. Don't you want to add a timeout liek after 30 seconds it will quit the loop no matter if it was finished or not?

AntonJansen replied:
Added a check. I think you have significant bigger issues when there is no port free anymore on your device :-). Anyway, I have added a timeout.

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/NetworkUtils.java:
```diff
@@ -0,0 +1,143 @@
+ */
+@NonNullByDefault
+public class NetworkUtils {
+
+    /**
+     * Check whether the host is available.
+     *
+     * @param host hostname or ip address of the host to check for availability
+     * @param timeout time in milliseconds for the check to timeout
+     * @param logger the logger to use for logging any issues
+     * @throws IOException if there is an unexpected error
+     * @throws BroadlinkHostNotReachableException when the device is not responding
+     */
+
+    public static void hostAvailabilityCheck(@Nullable String host, int timeout, Logger logger)
+            throws IOException, BroadlinkHostNotReachableException {
+        InetAddress address = InetAddress.getByName(host);
+        if (!address.isReachable(timeout)) {
+            throw new BroadlinkHostNotReachableException("Cannot reach " + host);
+        }
+    }
+
```
Method is no longer used
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,298 @@
+        this.thingConfig = getConfigAs(BroadlinkDeviceConfiguration.class);
+        // Validate whether the configuration makes any sense
+        if (thingConfig.isValidConfiguration().length() != 0) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
```
```suggestion
        if (!thingConfig.isValidConfiguration().isEmpty()) {
```

lsiepel commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -0,0 +1,298 @@
+        }
+
+        if (thingConfig.getPollingInterval() != 0) {
+            refreshHandle = scheduler.scheduleWithFixedDelay(new Runnable() {
+                @Override
+                public void run() {
+                    updateItemStatus();
+                }
+            }, 1L, thingConfig.getPollingInterval(), TimeUnit.SECONDS);
```
Suggestion might have whitespace issues. Meant as simplification.
```suggestion
            refreshHandle = scheduler.scheduleWithFixedDelay(this::updateItemStatus, 1L,
                    thingConfig.getPollingInterval(), TimeUnit.SECONDS);
```

lsiepel reviewed (changes requested):
Finished reviewing the last files, also found some other improvements. I think this is the last iteration, a lot of work has been done. Thank you for adapting the improvements regarding the excessive reachability checks. Much better now.

lsiepel commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,187 @@
+| mp2        | Broadlink MP2 WiFi Smart Power Strip (3 sockets, 3 USB)                       |
+| sp1        | Broadlink SP1 WiFi Smart Socket                                               |
+| sp2        | Broadlink SP2 WiFi Smart Socket with night light                              |
+| sp2s       | OEM SP2 Mini WiFi Smart Socket with night light                               |
+| sp3        | Broadlink SP3/Mini WiFi Smart Socket with night light                         |
+| sp3s       | Broadlink SP3s WiFi Smart Socket with Power Meter                             |
+| rm-pro     | Broadline RM Pro WiFi IR/RF Transmitter with temperature sensor               |
+| rm3        | Broadlink RM3/Mini WiFi IR Transmitter                                        |
+| rm3q       | Broadlink RM3 WiFi IR Transmitter with Firmware v44057                        |
+| rm4pro     | Broadlink RM4 Pro WiFi RF/IR Transmitter with temperature and humidity sensors|
+| rm4mini    | Broadlink RM4 mini WiFi IR Transmitter                                        |
```
JFTR: if changed, also i18n properties files, thing-type.xml and bindingconstant.java need to be adjusted

AntonJansen replied:
I suggest we break things now once and for all, before we have it published in the main distribution. We already have had several breaking changes in the channel names, so this one should just be another one.

AntonJansen commented on bundles/org.openhab.binding.broadlink/README.md:
```diff
@@ -0,0 +1,187 @@
+*-Remember if you manipulate the code file manually, remember to provide the proper location, and the proper ownership and permissions (Location is `$OPENHAB_USERDATA`, and permissions are `-rw-r--r-- 1 openhab openhab*
+
+## Full Example
+
```
I don't understand this comment. What are you expecting?

lsiepel replied:
Example for things and items are mandatory, sitemap is optional. 
The items example just has 1 item, that can be extended with some more. 
The things example is completely missing. Another exmaple from plugwiseha: https://www.openhab.org/addons/bindings/plugwiseha/#plugwiseha-things

AntonJansen replied:
Added an example of AC control using IR.

AntonJansen commented:
@lsiepel I resolved all the issues apart from one, for which I need a bit more input.

AntonJansen commented:
@lsiepel all known issues should now have been resolved. Let me know what to do next.

AntonJansen commented:
@Skinah All your requested changes should have been completed. Please confirm, so we can wrap-up this pull request.

lsiepel commented:
Thanks, now this is merged, you could add your binding's logo to the openHAB website. See https://next.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui
